### PR TITLE
Fix PHI non-null pattern

### LINKED ISSUE #104489: JIT: nullcheck is not folded for PHI
Seems like global assertprop bails out on the following snippet
```cs
bool Test(bool cond)
{
    object tmp;
    if (cond)
        tmp = "1111";
    else
        tmp = "2222";
    return tmp != null;
}
```
Current codegen:
```asm
; Assembly listing for method Proga:Test(ubyte):ubyte (FullOpts)
G_M26246_IG01:  ;; offset=0x0000
G_M26246_IG02:  ;; offset=0x0000
       mov      rax, 0x20000209798      ; '2222'
       mov      rdx, 0x200002097B8      ; '1111'
       test     cl, cl
       cmovne   rax, rdx
       test     rax, rax
       setne    al
       movzx    rax, al
G_M26246_IG03:  ;; offset=0x0023
       ret      
```
Expected codegen:
```asm
; Assembly listing for method Proga:Test(ubyte):ubyte (FullOpts)
G_M26246_IG01:  ;; offset=0x0000
G_M26246_IG02:  ;; offset=0x0000
       mov      eax, 1
G_M26246_IG03:  ;; offset=0x0023
       ret      
```

### PR DESCRIPTION
**Pull request by EgorBo**
Closes https://github.com/dotnet/runtime/issues/104489

Thanks to @AndyAyersMS and @jakobbotsch for ideas how to fix it.

Let's see what CI thinks..

### CODE REVIEW

**Comment by EgorBo** [comment]
@MihuBot

**Comment by EgorBo** [comment]
PTAL @AndyAyersMS @jakobbotsch cc @dotnet/jit-contrib [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=739004&view=ms.vss-build-web.run-extensions-tab)

I tried to fix the TP hit by only inspecting root node for PHI, but there is still 0.1% regression. Can we accept it?

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -1149,16 +1149,14 @@ AssertionIndex Compiler::optCreateAssertion(GenTree*         op1,
             }
         }
 
-        if (fgIsBigOffset(offset) || op1->gtOper != GT_LCL_VAR)
+        if (fgIsBigOffset(offset) || !op1->IsLocal())
```
Comment (jakobbotsch) [inline src/coreclr/jit/assertionprop.cpp:1152]: Is this change ok? `LCL_FLD` will create an assertion about the parent local without encoding any information about the offset, which seems odd.

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -11136,12 +11162,6 @@ void Compiler::fgValueNumberStore(GenTree* store)
 
             valueVNPair.SetBoth(initObjVN);
         }
-        else if (value->TypeGet() == TYP_REF)
-        {
-            // If we have an unsafe IL store of a TYP_REF to a non-ref (typically a TYP_BYREF)
-            // then don't propagate this ValueNumber to the lhs, instead create a new unique VN.
-            valueVNPair.SetBoth(vnStore->VNForExpr(compCurBB, store->TypeGet()));
-        }
```
Comment (jakobbotsch) [inline src/coreclr/jit/valuenum.cpp:11144]: Might be a good idea to see if you can dig up why this was added to verify it won't cause any problems to remove it.

**Comment by EgorBo** [comment]
superseded by https://github.com/dotnet/runtime/pull/108420

**Pull request closed without merging**